src/query/sql/src/planner/plans/hash_join.rs

```diff
@@ -92,13 +96,18 @@ impl PhysicalOperator for PhysicalHashJoin {
         {
             // TODO(leiysky): we can enforce redistribution here
             required.distribution = Distribution::Serial;
-        } else {
-            // A simple heuristic, we will always enforce `Hash` distribution for join
-            if child_index == 0 {
-                required.distribution = Distribution::Hash(self.probe_keys.clone());
+        } else if ctx.get_settings().get_prefer_broadcast_join()?
+            && !matches!(self.join_type, JoinType::Right | JoinType::Full)
```

Comment:
Disable broadcast join for right/full join.

Because `RightJoinCompactor` is executed on each node before executing `Merge Exchange`, it'll result in the wrong answer.

For example:

```sql
mysql> select * from t1;
+------+------+
| a    | b    |
+------+------+
|    1 |    2 |
|    3 |    4 |
|    7 |    8 |
+------+------+
3 rows in set (2.52 sec)
Read 3 rows, 24.00 B in 1.949 sec., 1.54 rows/sec., 12.32 B/sec.

mysql> select * from t2;
+------+------+
| a    | b    |
+------+------+
|    1 |    4 |
|    2 |    3 |
|    6 |    8 |
+------+------+
3 rows in set (2.39 sec)
Read 3 rows, 24.00 B in 1.833 sec., 1.64 rows/sec., 13.10 B/sec.

mysql> select * from t1 right join t2 on t1.a = t2.a;
+------+------+------+------+
| a    | b    | a    | b    |
+------+------+------+------+
| NULL | NULL |    1 |    4 |
| NULL | NULL |    2 |    3 |
| NULL | NULL |    6 |    8 |
|    1 |    2 |    1 |    4 |
| NULL | NULL |    2 |    3 |
| NULL | NULL |    6 |    8 |
| NULL | NULL |    1 |    4 |
| NULL | NULL |    2 |    3 |
| NULL | NULL |    6 |    8 |
+------+------+------+------+
9 rows in set (5.96 sec)
Read 6 rows, 48.00 B in 3.398 sec., 1.77 rows/sec., 14.13 B/sec.

```
t2 will be broadcasted to each node and `right join` with t1. Each node will produce three rows due to `RightJoinCompactor` :

```sql
| NULL | NULL |    1 |    4 |
| NULL | NULL |    2 |    3 |
| NULL | NULL |    6 |    8 |

|    1 |    2 |    1 |    4 |
| NULL | NULL |    2 |    3 |
| NULL | NULL |    6 |    8 |

| NULL | NULL |    1 |    4 |
| NULL | NULL |    2 |    3 |
| NULL | NULL |    6 |    8 |
```

I'm finding a good way to process the case, the intuitive way is to put `RightJoinCompactor` after `Merge Exchange`, but need to keep `join state` alive after join.

Reply:
https://github.com/datafuselabs/databend/pull/8811 ?

Reply:
> #8811 ?

I take a look at it, seems not related.